Remove unused locale messages

Gudahtt opened the pull request:
A number of unused locale messages have been removed - probably leftover from old UI elements that have since been removed.

The `verify_locale_strings` script has been augmented to search the UI for string literals, and match those against the locale message keys in the `en` locale. Any messages without a corresponding string literal are assumed to be unused.

The script has also been updated with an optional `--fix` parameter, which will automatically delete any unused messages from locales.

160 unused messages were found in this case, out of a total of about 650 messages. Another 70 messages are _potentially_ unused and require further investigation, but weren't as easy to rule out because they were found in string literals.

Gudahtt commented:
Here's the list of removed messages:
  - [ ] exposeAccounts
  - [ ] exposeDescription
  - [ ] confirmExpose
  - [ ] clearApprovalDataSuccess
  - [ ] aboutUs
  - [ ] accept
  - [ ] addCustomToken
  - [ ] addEthAddress
  - [ ] agreeTermsOfService
  - [ ] allDone
  - [ ] alreadyHaveSeedPhrase
  - [ ] amountPlusGas
  - [ ] amountPlusTxFee
  - [ ] appDescription
  - [ ] available
  - [ ] balances
  - [ ] beta
  - [ ] borrowDharma
  - [ ] buy
  - [ ] buyCoinbase
  - [ ] buyCoinbaseExplainer
  - [ ] cancelN
  - [ ] classicInterface
  - [ ] clickCopy
  - [ ] company
  - [ ] confirmationTime
  - [ ] confirmContract
  - [ ] confirmTransaction
  - [ ] connecting
  - [ ] connectingToUnknown
  - [ ] connectToLedger
  - [ ] connectToTrezor
  - [ ] continue
  - [ ] continueToCoinbase
  - [ ] copiedClipboard
  - [ ] copiedSafe
  - [ ] createDen
  - [ ] crypto
  - [ ] currentNetwork
  - [ ] currentRpc
  - [ ] customize
  - [ ] denExplainer
  - [ ] depositBTC
  - [ ] depositEth
  - [ ] depositFiat
  - [ ] depositFromAccount
  - [ ] depositShapeShift
  - [ ] depositShapeShiftExplainer
  - [ ] directDeposit
  - [ ] dontHaveAHardwareWallet
  - [ ] editNetwork
  - [ ] editAccountName
  - [ ] encryptNewDen
  - [ ] ensNameNotFound
  - [ ] enterPasswordConfirm
  - [ ] eth
  - [ ] exchangeRate
  - [ ] exportPrivateKeyWarning
  - [ ] fastest
  - [ ] feeChartTitle
  - [ ] followTwitter
  - [ ] fromToSame
  - [ ] gasLimitRequired
  - [ ] generatingSeed
  - [ ] gasPriceRequired
  - [ ] generatingTransaction
  - [ ] hardwareWallets
  - [ ] hardwareWalletsMsg
  - [ ] havingTroubleConnecting
  - [ ] hereList
  - [ ] howToDeposit
  - [ ] holdEther
  - [ ] importAnAccount
  - [ ] importDen
  - [ ] importWithSeedPhrase
  - [ ] invalidGasParams
  - [ ] invalidRequest
  - [ ] jsonFail
  - [ ] keepTrackTokens
  - [ ] knowledgeDataBase
  - [ ] legal
  - [ ] limit
  - [ ] loose
  - [ ] loweCaseWords
  - [ ] min
  - [ ] minutesShorthand
  - [ ] mobileSyncTitle
  - [ ] needImportPassword
  - [ ] newPassword8Chars
  - [ ] newRecipient
  - [ ] showAdvancedOptions
  - [ ] hideAdvancedOptions
  - [ ] optionalNickname
  - [ ] noTransactionHistory
  - [ ] notFound
  - [ ] notStarted
  - [ ] oldUI
  - [ ] oldUIMessage
  - [ ] openInTab
  - [ ] or
  - [ ] outgoing
  - [ ] originalTotal
  - [ ] passwordCorrect
  - [ ] passwordMismatch
  - [ ] passwordShort
  - [ ] pleaseReviewTransaction
  - [ ] popularTokens
  - [ ] readMore
  - [ ] readMore2
  - [ ] receive
  - [ ] refundAddress
  - [ ] retryWithMoreGas
  - [ ] readyToConnect
  - [ ] sampleAccountName
  - [ ] saveAsFile
  - [ ] saveSeedAsFile
  - [ ] secondsShorthand
  - [ ] select
  - [ ] selectService
  - [ ] selectAnAddress
  - [ ] selectAnAsset
  - [ ] sendTokensAnywhere
  - [ ] shapeshiftBuy
  - [ ] showQRCode
  - [ ] signMessage
  - [ ] sigRequested
  - [ ] spaceBetween
  - [ ] speedUpTitle
  - [ ] speedUpSubtitle
  - [ ] step1HardwareWallet
  - [ ] step1HardwareWalletMsg
  - [ ] step2HardwareWallet
  - [ ] step2HardwareWalletMsg
  - [ ] step3HardwareWallet
  - [ ] step3HardwareWalletMsg
  - [ ] takesTooLong
  - [ ] tokenAddress
  - [ ] tokenSelection
  - [ ] tokenWarning1
  - [ ] transactionWithNonce
  - [ ] transactionUpdatedGas
  - [ ] transactionMemo
  - [ ] transactionNumber
  - [ ] transfers
  - [ ] trezorHardwareWallet
  - [ ] twelveWords
  - [ ] uiMigrationAnnouncement
  - [ ] uiWelcome
  - [ ] uiWelcomeMessage
  - [ ] unavailable
  - [ ] unknownNetworkId
  - [ ] usaOnly
  - [ ] useOldUI
  - [ ] vaultCreated
  - [ ] viewNetworkInfo
  - [ ] whatsThis
  - [ ] yourUniqueAccountImage
  - [ ] yourUniqueAccountImageDescription1
  - [ ] yourUniqueAccountImageDescription2
  - [ ] yourUniqueAccountImageDescription3

whymarrh commented on development/verify-locale-strings.js:
```diff
@@ -4,37 +4,87 @@
+// A report will be printed to the console detailing any unused locales, and also
+// any missing messages in the non-English locales.
+//
+// The if theoptional '--fix' parameter is given, locales will be automatically
```
```suggestion
// The if the optional '--fix' parameter is given, locales will be automatically
```

Gudahtt replied:
Thanks, done.

whymarrh commented on development/verify-locale-strings.js:
```diff
@@ -4,37 +4,87 @@
+//     node app/scripts/verify-locale-strings.js [<locale>] [--fix]
 //
-// will check the given locale against the strings in english
+// This script will validate that no locales have unused messages. It will check
```
```suggestion
// This script will validate that locales have no unused messages. It will check
```

Gudahtt replied:
Thanks, done

Gudahtt commented:
I've found these additional messages that have corresponding string literals, but seem to be unused:
Messages not present in UI:
  - [ ] add
  - [ ] address
  - [ ] addressBook
  - [ ] editingTransaction
  - [ ] gas
  - [ ] gasFee
  - [ ] info
  - [ ] onlySendToEtherAddress
  - [ ] onlySendTokensToAccountAddress
  - [ ] restoreVault
  - [ ] rpc
  - [ ] status
  - [ ] tokenBalance
  - [ ] transactions
  - [ ] warning

Gudahtt committed:
Switch to using string literals for locale keys

Various message keys were being specified with a string template
instead of a string literal. They have been switched to use string
literals so that the script for detecting unused messages can find
them.

Gudahtt commented:
I've just updated this to restore a few messages on the `connect-hardware` screen that were using string templates for the message key. These are the restored messages:
 - [ ] dontHaveAHardwareWallet
 - [ ] hardwareWallets
 - [ ] hardwareWalletsMsg
 - [ ] havingTroubleConnecting
 - [ ] readyToConnect
 - [ ] step1HardwareWallet
 - [ ] step1HardwareWalletMsg
 - [ ] step2HardwareWallet
 - [ ] step2HardwareWalletMsg
 - [ ] step3HardwareWallet
 - [ ] step3HardwareWalletMsg

Gudahtt committed:
Remove additional unused locale messages

The following messages were more difficult to rule out because they
were present as string literals in the UI. They do appear to be
unused as locale keys though.

Here are the removed messages:
  - [ ] add
  - [ ] address
  - [ ] addressBook
  - [ ] editingTransaction
  - [ ] gas
  - [ ] gasFee
  - [ ] info
  - [ ] onlySendToEtherAddress
  - [ ] onlySendTokensToAccountAddress
  - [ ] restoreVault
  - [ ] rpc
  - [ ] status
  - [ ] tokenBalance
  - [ ] transactions
  - [ ] warning

Gudahtt committed:
Remove unused locale messages

A number of unused locale messages have been removed - probably
leftover from old UI elements that have since been removed.

The `verify_locale_strings` script has been augmented to search the UI
for string literals, and match those against the locale message keys in
the `en` locale. Any messages without a corresponding string literal
are assumed to be unused.

The script has also been updated with an optional `--fix` parameter,
which will automatically delete any unused messages from locales.

148 unused messages were found in this case, out of a total of about
650 messages. Another 70 messages are _potentially_ unused and require
further investigation, but weren't as easy to rule out because they
were found in string literals.

Gudahtt commented:
This should be in a decent state to review now. I'm _reasonably_ confident that everything in the first batch of messages I removed is truly unused. The second batch I'm less certain about - I certainly can't find them used anywhere, and I haven't noticed any console warnings about them being missing when using the app, but they might be worth some additional scrutiny.

They're listed here: https://github.com/MetaMask/metamask-extension/pull/7190#issuecomment-532858423

whymarrh reviewed (approved):
Seems legit

danfinlay commented on development/verify-locale-strings.js:
```diff
@@ -4,37 +4,87 @@
+//     node app/scripts/verify-locale-strings.js [<locale>] [--fix]
 //
-// will check the given locale against the strings in english
+// This script will validate that locales have no unused messages. It will check
```
It sounds like we might want the inverse also, based on the concerns you've raised: Validate that no UI template call uses a non-existent message. Not sure how efficiently this can be done, but if we assume the template string is always in a `t()` call, it might be efficient enough to regex.

Just nervous about removing used messages, like you suggested might be the case!

Gudahtt replied:
I tried this already but too many cases use variables instead of string literals.

Many of those cases could be eliminated pretty easy, but there are a few that are very dynamic that would be challenging to refactor.

A better way to catch these cases might be to send an error to Sentry when we detect a missing key from the `en` locale, rather than just print a console warning like we do now.

danfinlay replied:
Yeah, that'd be a good use of sentry.

danfinlay reviewed (approved):
Coupled with some sentry reporting in the case of a missing string, this should be fine. That's a lot of removed dead lines!
